### PR TITLE
feat: Add VM commands

### LINKED ISSUE #5: Support for basic VM actions
Support:
- Listing VMs for an organisation
- Shutdown
- Start
- Stop
- Reset

VM actions should be able to use the ID or the FQDN of the VM.

### PR DESCRIPTION
**Pull request by JakeMakesStuff**
Fixes #5

### CODE REVIEW

File: cmd/katapult/vms.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+				pagination := resp.Pagination
+				page = pagination.CurrentPage + 1
+				perPage = pagination.PerPage
+				end := pagination.TotalPages == 0 || pagination.CurrentPage == pagination.TotalPages
+
+				for _, vm := range vms {
+					fqdn := vm.FQDN
+					if fqdn == "" {
+						fqdn = "<no fqdn specified>"
+					}
```
Comment (jimeh) [inline cmd/katapult/vms.go:104]: I don't think this check is needed. As far as I know, VMs will always have a FDQN value.

File: cmd/katapult/vms_test.go [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+	switch {
+	case ref.ID != "":
+		delete(v.powerStates, "i"+ref.ID)
+	case ref.FQDN != "":
+		delete(v.powerStates, "s"+ref.FQDN)
+	}
+	return nil, nil, nil
+}
+
+func TestVMs_List(t *testing.T) {
```
Comment (jimeh) [inline cmd/katapult/vms_test.go:157]: There's no test case for what happens if no org ID or sub-domain is provided.

File: cmd/katapult/vms_test.go [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+			cmd := virtualMachinesCmd(&vmsClient{organizationIDPages: tt.id, organizationSubdomainPages: tt.subdomains})
+			cmd.SetArgs(tt.args)
+			assertCobraCommand(t, cmd, tt.wantErr, tt.want, tt.stderr)
+		})
+	}
+}
+
+func TestVMs_Poweroff(t *testing.T) {
+	successMessage := "Virtual machine successfully powered down.\n"
+	tests := []struct {
```
Comment (jimeh) [inline cmd/katapult/vms_test.go:267]: There's no test case for when no VM ID or FQDN is provided. The other power-state tests are also missing such a test case.

**Review by jimeh — CHANGES_REQUESTED** [review]

File: cmd/katapult/vms.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+		Long:    "Get a list of virtual machines from an organization.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ref := core.OrganizationRef{ID: cmd.Flags().Lookup("id").Value.String()}
+			if ref.ID == "" {
+				subdomain := cmd.Flags().Lookup("subdomain").Value.String()
+				if subdomain == "" {
+					return fmt.Errorf("both ID and subdomain are unset")
+				}
+				ref.SubDomain = subdomain
+			}
```
[the remark this answers is not in the document]
Comment (JakeMakesStuff) [reply cmd/katapult/vms.go]: > Ah, not to mention the Org ID / SubDomain should probably be read from the config / environment vars too, meaning it defaults to what's pulled out of config, and the arg/flag lets you override it, or specify it if you don't have one configured.

I have went ahead and made an issue for this. 👍

File: cmd/katapult/vms.go [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+	start := &cobra.Command{
+		Use:   "start",
+		Short: "Used to start a virtual machine.",
+		Long:  "Used to start a virtual machine.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ref, err := vmFlag(cmd)
+			if err != nil {
+				return err
+			}
+			_, resp, err := client.Start(cmd.Context(), ref)
```
[the remark this answers is not in the document]
Comment (JakeMakesStuff) [reply cmd/katapult/vms.go]: > Out of scope for this PR I think, so I'm just mentioning it for future reference;
> 
> We will eventually need a `--wait` flag or something similar on these power state commands, to have them wait for the returned task object to be marked as completed. Effectively blocking until the operation has completed and the VM is in the new and expected state.

I have went ahead and made an issue for this. 👍

**Review by strideynet — APPROVED** [review]

**Review by jimeh — CHANGES_REQUESTED** [review]
This is looking good, only thing missing (which I missed myself last time around 🤦‍♂️), are test cases for the power state commands for when a VM does not exist.

File: cmd/katapult/vms_test.go [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+		CurrentPage: opts.Page,
+		TotalPages:  len(pages),
+		Total:       totalItems,
+		PerPage:     len(page),
+	}}, nil
+}
+
+func (v *vmsClient) ensureFound(ref core.VirtualMachineRef) error {
+	if (ref.FQDN != "" && v.fqdnNotFound == ref.FQDN) || (ref.ID != "" && v.idNotFound == ref.ID) {
+		return katapult.ErrNotFound
```
Comment (jimeh) [inline cmd/katapult/vms_test.go:84]: This should be a `core.ErrVirtualMachineNotFound` to match what the go-katapult package returns when looking up Virtual Machines that don't exist.

```suggestion
		return core.ErrVirtualMachineNotFound
```

The error chain for `core.ErrVirtualMachineNotFound` is: 

- `katapult.Err`
  - `katapult.ErrResponse`
    - `katapult.ErrNotFound`
      - `katapult.ErrResourceNotFound`
        - `core.ErrVirtualMachineNotFound`

It's a bit long, but it ensures that it's easy to target error handling at any specific level you might care about.

**Review by jimeh — CHANGES_REQUESTED** [review]

**Review by jimeh — APPROVED** [review]
👍

**Pull request merged**